src/Strategy/Python/Poetry/PoetryLock.hs

```diff
@@ -0,0 +1,168 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Strategy.Python.Poetry.PoetryLock (
+  PoetryLock (..),
+  PoetryMetadata (..),
+  PoetryLockPackage (..),
+  PackageName (..),
+  PoetryLockPackageSource (..),
+  PoetryLockDependencySpec (..),
+  ObjectVersion (..),
+  poetryLockCodec,
+  toMap,
+) where
+
+import Control.Applicative (Alternative ((<|>)))
+import Data.Map (Map)
+import Data.Map.Strict qualified as M
+import Data.Text (Text, toLower)
+import DepTypes
+import Toml (TomlCodec, (.=))
+import Toml qualified
+
+-- | Content of poetry lock file
+data PoetryLock = PoetryLock
+  { poetryLockPackages :: [PoetryLockPackage]
+  , poetryLockMetadata :: !PoetryMetadata
+  }
+  deriving (Eq, Ord, Show)
+
+newtype PackageName = PackageName {unPackageName :: Text} deriving (Eq, Ord, Show)
+
+poetryLockCodec :: TomlCodec PoetryLock
+poetryLockCodec =
+  PoetryLock
+    <$> Toml.list poetryLockPackageCodec "package" .= poetryLockPackages
+    <*> Toml.table poetryMetadataCodec "metadata" .= poetryLockMetadata
+
+-- | Metadata of poetry lock file
+data PoetryMetadata = PoetryMetadata
+  { poetryMetadataLockVersion :: Text
+  , poetryMetadataContentHash :: Text
+  , poetryMetadataPythonVersions :: Text
+  }
+  deriving (Eq, Ord, Show)
+
+poetryMetadataCodec :: TomlCodec PoetryMetadata
+poetryMetadataCodec =
+  PoetryMetadata
+    <$> Toml.diwrap (Toml.text "lock-version") .= poetryMetadataLockVersion
+    <*> Toml.diwrap (Toml.text "content-hash") .= poetryMetadataContentHash
+    <*> Toml.diwrap (Toml.text "python-versions") .= poetryMetadataPythonVersions
+
+-- | Source of poetry pacakge found in lock file.
+data PoetryLockPackageSource = PoetryLockPackageSource
+  { poetryLockPackageSourceType :: Text
+  , poetryLockPackageSourceUrl :: Text
+  , poetryLockPackageSourceReference :: Maybe Text
+  , poetryLockPackageSourceResolvedReference :: Maybe Text
+  }
+  deriving (Eq, Ord, Show)
+
+-- | Poetry package entry found in poetry lock file
+data PoetryLockPackage = PoetryLockPackage
+  { poetryLockPackageName :: PackageName
+  , poetryLockPackageVersion :: Text
+  , poetryLockPackageCategory :: Text
+  , poetryLockPackageOptional :: Bool
+  , poetryLockPackagePythonVersions :: Text
+  , poetryLockPackageDependencies :: Map Text PoetryLockDependencySpec
+  , poetryLockPackageSource :: Maybe PoetryLockPackageSource
+  }
+  deriving (Eq, Ord, Show)
+
+poetryLockPackageCodec :: TomlCodec PoetryLockPackage
+poetryLockPackageCodec =
+  PoetryLockPackage
+    <$> Toml.diwrap (Toml.text "name") .= poetryLockPackageName
```

Comment:
[no action]: THIS is the correct use of `diwrap`, since we have to `coerce` a `Text` into a `PackageName`.  The reason that works is this:

```haskell
-- `a ~R# b` means "`a` has the same runtime representation as `b`"
class a ~R# b => Coercible (a :: k) (b :: k)

-- In docs, this has a complex type that means the same as `a ~R# b` above.
coerce :: Coercible a b => a -> b

-- Simple, mostly useless data types, but could be useful, like `Text` and `PackageName`, for instance.
data NotSpecial = NotSpecial
newtype Special = Special { notSpecial :: NotSpecial }

-- This typechecks, explanation below
makeSpecial :: NotSpecial -> Special
makeSpecial = coerce
```

During typechecking, GHC will detect that the runtime repr for these types (`Special` and `NotSpecial`) are the same, and will automatically derive a `Coercible` instance between the two (in both directions).  `diwrap` uses this to magically move `NotSpecial` types into `Special` types.  Or in this case, lifting `Text` into `PackageName`.

Reply:
This is the recommended `newtype` handling method for this library.  I don't love it, but it works unambiguously and safely, since the newtype deriving rules are compiler-guaranteed.

As for using `fmap FooConstructor`, I think it's less readable to use a separate functor inside of the top-level applicative that's constructing the record type that the newtype lives in.  To avoid using `coerce`, and also avoid nested `fmap`-ping, we need a `Toml.diwrap`-like helper that takes an explicit constructor.